examples/stream-to-file/src/main.rs

```diff
@@ -0,0 +1,115 @@
+//! Run with
+//!
+//! ```not_rust
+//! cargo run -p example-stream-to-file
+//! ```
+
+use axum::{
+    body::Bytes,
+    extract::{BodyStream, Multipart, Path},
+    http::StatusCode,
+    response::{Html, Redirect},
+    routing::{get, post},
+    BoxError, Router,
+};
+use futures::{Stream, TryStreamExt};
+use std::{io, net::SocketAddr};
+use tokio::{fs::File, io::BufWriter};
+use tokio_util::io::StreamReader;
+use tracing_subscriber::{layer::SubscriberExt, util::SubscriberInitExt};
+
+#[tokio::main]
+async fn main() {
+    tracing_subscriber::registry()
+        .with(tracing_subscriber::EnvFilter::new(
+            std::env::var("RUST_LOG").unwrap_or_else(|_| "example_stream_to_file=debug".into()),
+        ))
+        .with(tracing_subscriber::fmt::layer())
+        .init();
+
+    let app = Router::new()
+        .route("/", get(show_form).post(accept_form))
+        .route("/file/:file_name", post(save_request_body));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+// Handler that streams the request body to a file.
+//
+// POST'ing to `/file/foo.txt` will create a file called `foo.txt`.
+async fn save_request_body(
+    Path(file_name): Path<String>,
+    body: BodyStream,
+) -> Result<(), (StatusCode, String)> {
+    stream_to_file(&file_name, body)
+        .await
+        .map_err(|err| (StatusCode::INTERNAL_SERVER_ERROR, err.to_string()))
+}
+
+// Handler that returns HTML for a multipart form.
+async fn show_form() -> Html<&'static str> {
+    Html(
+        r#"
+        <!doctype html>
+        <html>
+            <head></head>
+            <body>
+                <form action="/" method="post" enctype="multipart/form-data">
+                    <div>
+                        <label>
+                            Upload file:
+                            <input type="file" name="file" multiple>
+                        </label>
+                    </div>
+
+                    <div>
+                        <input type="submit" value="Upload files">
+                    </div>
+                </form>
+            </body>
+        </html>
+        "#,
+    )
+}
+
+// Handler that accepts a multipart form upload and streams each field to a file.
+async fn accept_form(mut multipart: Multipart) -> Result<Redirect, (StatusCode, String)> {
+    while let Some(field) = multipart.next_field().await.unwrap() {
+        let file_name = if let Some(file_name) = field.file_name() {
+            file_name.to_owned()
```

Comment:
Is the `.to_owned()` necessary?

Reply:
Yeah because we move `field` into `stream_to_file` while `file_name` is borrowed from `field`:

```
error[E0505]: cannot move out of `field` because it is borrowed
  --> examples/stream-to-file/src/main.rs:91:35
   |
85 |         let file_name = if let Some(file_name) = field.file_name() {
   |                                                  ----------------- borrow of `field` occurs here
...
91 |         stream_to_file(file_name, field)
   |         --------------            ^^^^^ move out of `field` occurs here
   |         |
   |         borrow later used by call
```